aws-rds-dbinstance/src/main/java/software/amazon/rds/dbinstance/BaseHandlerStd.java

```diff
@@ -715,150 +393,48 @@ private void resourceStabilizationTime(final CallbackContext context) {
-                        DomainMembershipStatus.KerberosEnabled.equalsString(membership.status()));
-    }
-
-    boolean isVpcSecurityGroupsActive(final DBInstance dbInstance) {
-        return Optional.ofNullable(dbInstance.vpcSecurityGroups()).orElse(Collections.emptyList())
-                .stream()
-                .allMatch(group -> VPCSecurityGroupStatus.Active.equalsString(group.status()));
-    }
-
-    boolean isNoPendingChanges(final DBInstance dbInstance) {
-        final PendingModifiedValues pending = dbInstance.pendingModifiedValues();
-        return (pending == null) || (pending.dbInstanceClass() == null &&
-                pending.allocatedStorage() == null &&
-                pending.automationMode() == null &&
-                pending.backupRetentionPeriod() == null &&
-                pending.dbInstanceIdentifier() == null &&
-                pending.dbSubnetGroupName() == null &&
-                pending.engine() == null &&
-                pending.engineVersion() == null &&
-                pending.iamDatabaseAuthenticationEnabled() == null &&
-                pending.iops() == null &&
-                pending.licenseModel() == null &&
-                pending.masterUserPassword() == null &&
-                pending.multiAZ() == null &&
-                pending.pendingCloudwatchLogsExports() == null &&
-                pending.port() == null &&
-                CollectionUtils.isNullOrEmpty(pending.processorFeatures()) &&
-                pending.resumeFullAutomationModeTime() == null &&
-                pending.storageThroughput() == null &&
-                pending.storageType() == null
-        );
-    }
-
-    boolean isCaCertificateChangesApplied(final DBInstance dbInstance, final ResourceModel model) {
-        final PendingModifiedValues pending = dbInstance.pendingModifiedValues();
-        return pending == null ||
-                pending.caCertificateIdentifier() == null ||
-                BooleanUtils.isNotTrue(model.getCertificateRotationRestart());
-    }
-
-    boolean isDBParameterGroupNotApplying(final DBInstance dbInstance) {
-        return Optional.ofNullable(dbInstance.dbParameterGroups()).orElse(Collections.emptyList())
-                .stream()
-                .noneMatch(group -> DBParameterGroupStatus.Applying.equalsString(group.parameterApplyStatus()));
-    }
-
-    boolean isReplicationComplete(final DBInstance dbInstance) {
-        return Optional.ofNullable(dbInstance.statusInfos()).orElse(Collections.emptyList())
-                .stream()
-                .filter(statusInfo -> READ_REPLICA_STATUS_TYPE.equals(statusInfo.statusType()))
-                .allMatch(statusInfo -> ReadReplicaStatus.Replicating.equalsString(statusInfo.status()));
+        final Optional<DBCluster>  maybeDbCluster = DBInstancePredicates.isDBClusterMember(model)
+            ? Optional.of(fetchDBCluster(rdsProxyClient, model))
+            : Optional.empty();
```

Comment:
Can we not use optionals, it adds more cognitive load for the reader, can we instead use method overloads for when there is a dbcluster vs when there isn't?

From Brian Goetz himself: https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type/26328555#26328555

And: http://dolszewski.com/java/java-8-optional-use-cases/

Reply:
We can not-use Optionals here. But let's avoid taking the java authors word as gospel. What they intended Optionals for says nothing about what they are good for. What they are good for is avoiding `null`. The cognitive load of an optional is really just the cognitive load of a nullable reference, but it's in your face so you/we are more likely to handle it correctly.